### PR TITLE
signing implementation and arch variants

### PR DESCRIPTION
**Pull request by TheJJ**
- `debmagic sign` to sign `.changes`
- add `host_arch_variant` building for amd64v3 architecture variant of amd64
- color printing support of progress
- default container driver selection in config

### CODE REVIEW

File: packages/debmagic/src/driver/driver_docker.rs [hunk 1]
```diff
@@ -462,59 +463,149 @@ impl Driver for DriverDocker {
     fn isolation_capability(&self) -> IsolationCapability {
         IsolationCapability::Container
     }
+
+    fn sign_changes(&self, request: &SignRequest) -> anyhow::Result<()> {
```
Comment (mikonse) [inline packages/debmagic/src/driver/driver_docker.rs:459]: Overall I'm not sure if the sign methods and implementations should really be handled by the drivers themselves. We can already see that this leads to some weird case handling and duplications between the ephemeral and build container handling which apparently requires separate implementations. I get why the current implementation puts it in the driver as otherwise one cannot reuse a build container for signing as when starting the container we don't yet know that we e.g. need to map the gpg socket and copy the respective stuff into the container.
I'm just wondering if there would be a better abstraction here s.t. we can just have the signing workflow as another plain consumer of the driver interface like test and build and keep the special handling inside the driver to a minimum.
But thats maybe something for a future refactoring

File: packages/debmagic/src/main.rs [hunk 2]
```diff
@@ -51,11 +53,21 @@ fn run() -> anyhow::Result<ExitCode> {
                 BuildTarget::Source(source_args) => (&source_args.build, None, None, true),
             };
 
+            let config_driver = load_config(
```
Comment (mikonse) [inline packages/debmagic/src/main.rs]: this looks somewhat as if we will be loading the config twice now. Initially here and then another time down the line when e.g. resolving the settings for the drivers themselves.

Probably the BuildIntentInput should already receive a fully resolved config object instead of the config file path (other subcommands should be handled similarly). That way we can avoid that duplication where we might resolve configs differently in different places.

File: packages/debmagic/src/cli.rs [hunk 3]
```diff
@@ -107,43 +137,81 @@ pub struct CommonBuildArgs {
...
     #[arg(
         long,
-        action = clap::ArgAction::SetFalse,
-        help = "Do not sign the resulting .changes/.dsc, overriding a 'sign_package = true' default in the config file."
+        num_args = 0..=1,
+        default_missing_value = "true",
+        action = clap::ArgAction::Set,
+        help = "Do not sign the resulting .changes/.dsc, overriding a 'sign.source = true' default in the config file."
     )]
     pub no_sign: Option<bool>,
```
Comment (mikonse) [inline packages/debmagic/src/cli.rs]: Defining these cli options this way looks a bit weird, is there no other way with clap creating these --no-sign / --sign pairs automatically?

Also we should not be passing both flags to further program stages as that is a very weird api. This combination is something specific to the cli. Any signing logic should just see one boolean, whether to sing or not to sign.

File: packages/debmagic/src/cli.rs [hunk 4]
```diff
@@ -107,43 +137,81 @@ pub struct CommonBuildArgs {
...
+    pub sign_notify: Option<bool>,
+
+    #[arg(
+        long = "no-sign-notify",
+        num_args = 0..=1,
+        default_missing_value = "true",
+        action = clap::ArgAction::Set,
+        help = "Do not send a signing notification, overriding a 'sign.notify = true' default in the config file."
+    )]
+    pub no_sign_notify: Option<bool>,
```
Comment (mikonse) [inline packages/debmagic/src/cli.rs]: See comment regarding --sign / --no-sign weirdness

File: packages/debmagic/src/cli.rs [hunk 5]
```diff
@@ -183,12 +251,21 @@ pub struct BinaryTargetArgs {
...
+        action = clap::ArgAction::Set,
+        help = "Synchronize changed source inputs while preserving build outputs. Implies --persistent"
+    )]
     pub incremental: Option<bool>,
 
     #[arg(
         long = "debug-symbols",
-        action = clap::ArgAction::SetTrue,
+        num_args = 0..=1,
+        default_missing_value = "true",
```
Comment (mikonse) [inline packages/debmagic/src/cli.rs:302]: are you sure the default missing value does what we want?

Choosing an `Option` type here was done explicitly so we can differentiate between that flag (especially for the --xxx / --no-xxx flag pairs) was passed on the command line or wasn't passed so we can properly apply overriding chains with our chained config loading logic.

File: packages/debmagic/src/driver/driver_docker.rs (hunk 1 above)
Comment (TheJJ) [reply packages/debmagic/src/driver/driver_docker.rs:459]: best solution, i think: just implement signing on the host in debmagic, instead of the glorious 700loc shell script of debsign today.

File: packages/debmagic/src/cli.rs (hunk 3 above)
Comment (TheJJ) [reply packages/debmagic/src/cli.rs]: i tried, but i couldn't find a better way to have tristate flags: no value, and 2 override values - no value is used so we can keep the config setting. also this supports sign=false, which for me is more intuitive than no-sign

File: packages/debmagic/src/cli.rs (hunk 5 above)
Comment (TheJJ) [reply packages/debmagic/src/cli.rs:302]: see above

File: packages/debmagic/src/cli.rs (hunk 3 above)
Comment (TheJJ) [reply packages/debmagic/src/cli.rs]: fixed, we removed the `--no-*` arguments in favor of the fancy bool parsing that accepts yes,true,1/no,false,0 etc.

File: packages/debmagic/src/main.rs (hunk 2 above)
Comment (TheJJ) [reply packages/debmagic/src/main.rs]: gone when we removed the in-container signing approach

**Review by mikonse — APPROVED** [review]
